### ISSUE TITLE
Swift access race when Thread Sanitizer turned on

### ISSUE THREAD

**Issue by strong84** [dropbox/SwiftyDropbox#422]
If Xcode Thread Sanitizer turned on, I got many Data Race warnings when doing uploads:
`
SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:367 Swift access race in SwiftyDropbox.RequestWithTokenRefresh.StateAccess.accessStateWithLock<τ_0_0>(block: (SwiftyDropbox.RequestWithTokenRefresh.MutableState) throws -> τ_0_0) throws -> τ_0_0 at 0x12af06b00`

warnings generated from here:
https://github.com/dropbox/SwiftyDropbox/blob/aee63b4717fda3255c532d52a4b05f0328ac00bc/Source/SwiftyDropbox/Shared/Handwritten/Request%2BTokenRefresh.swift#L79

Although the upload is successful, But I noticed some failures when uploading multiple files.

**Comment by strong84** [comment]
`/SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:367 Swift access race in SwiftyDropbox.RequestWithTokenRefresh.StateAccess.accessStateWithLock<τ_0_0>(block: (SwiftyDropbox.RequestWithTokenRefresh.MutableState) throws -> τ_0_0) throws -> τ_0_0 at 0x12af06b00

Location is a 28-byte heap object at 0x12af06b00
Mutating access by thread 4
#0	0x00000001078a700c in RequestWithTokenRefresh.StateAccess.accessStateWithLock<τ_0_0>(block:) ()
#1	0x00000001078b1c44 in RequestWithTokenRefresh.accessStateWithLock<τ_0_0>(block:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:367
#2	0x00000001078b0904 in RequestWithTokenRefresh.handleTokenRefreshResult(_:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:332
#3	0x00000001078ae458 in closure #1 in closure #2 in RequestWithTokenRefresh.init(requestCreation:onTaskCreation:authStrategy:filesAccess:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:244
#7	0x000000010b38173c in _dispatch_client_callout ()
Mutating access by thread 5
#0	0x00000001078a701c in RequestWithTokenRefresh.StateAccess.accessStateWithLock<τ_0_0>(block:) ()
#1	0x00000001078b1c44 in RequestWithTokenRefresh.accessStateWithLock<τ_0_0>(block:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:367
#2	0x00000001078aea24 in RequestWithTokenRefresh.setCompletionHandler(queue:completionHandler:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:270
#4	0x00000001078bc3d8 in UploadRequest.response(queue:completionHandler:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request.swift:129
#6	0x00000001078a4fc0 in closure #1 in HasRequestResponse.responseResult() at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+Async.swift:26
#8	0x000000020a881744 in closure #1 in withCheckedContinuation<τ_0_0>(function:_:) ()
Heap block allocated by thread 5
#0	0x000000010c3b3a4c in wrap_malloc ()
#1	0x00000001801c33f0 in _malloc_type_malloc_outlined ()
#2	0x00000001078ad6c8 in RequestWithTokenRefresh.init(requestCreation:onTaskCreation:authStrategy:filesAccess:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/Request+TokenRefresh.swift:227
#4	0x000000010780dfc0 in closure #4 in DropboxTransportClientImpl.init(authStrategy:baseHosts:userAgent:firstPartyUserAgent:selectUser:sessionConfiguration:sessionCreation:longpollSessionConfiguration:longpollSessionCreation:filesAccess:authChallengeHandler:pathRoot:headersForRouteHost:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/DropboxTransportClient.swift:232
#6	0x0000000107855ec8 in closure #1 in NetworkSessionManager.registeredApiRequest(from:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/NetworkSession/NetworkSessionManager.swift:171
#10	0x00000001c5bfb7c4 in implicit closure #2 in implicit closure #1 in OS_dispatch_queue.asyncAndWait<τ_0_0>(execute:) ()
#11	0x0000000107855108 in NetworkSessionManager.apiRequestUpload(request:input:networkTaskTag:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/NetworkSession/NetworkSessionManager.swift:133
#12	0x000000010780f760 in DropboxTransportClientImpl.request<τ_0_0, τ_0_1, τ_0_2>(_:serverArgs:input:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Handwritten/DropboxTransportClient.swift:273
#14	0x0000000106959474 in FilesRoutes.upload(path:mode:autorename:clientModified:mute:propertyGroups:strictConflict:contentHash:input:) at /SwiftyDropbox/Source/SwiftyDropbox/Shared/Generated/FilesRoutes.swift:1741
#15	0x00000001051153e8 in closure #1 in static Dropbox.upload(filename:data:_:) at /App/Dropbox.swift:202
#16	0x000000020a8b410c in swift::runJobInEstablishedExecutorContext(swift::Job*) ()`

**Comment by greg-db** [comment]
We'll be happy to look into this, but we could use some more information. Please reply with:

- the version numbers of the platform and SDK you are using
- the steps to reproduce the issue, including relevant code snippet(s), but don't include any access or refresh token(s)

**Comment by dsmurfin** [comment]
I am also seeing this with:

- iOS (17.5), SDK 10.0.7

The simplest reproducer is to call the function below, for a linked client.

```swift
func getAccount() {
    if let currentAccount = DropboxClientsManager.authorizedClient?.users.getCurrentAccount() {
        Task {
            do {
                let result = try await currentAccount.response()
                print(result)
            } catch {
                print(error)
            }
        }
    }
}

```

**Comment by greg-db** [comment]
Thanks! I'll ask the team to look into this.

**Comment by greg-db** [comment]
This should be fixed in v10.1.0. Please try that out and let us know if you're still seeing any issues. Thanks!
